security/medusa/l3/registry.c

```diff
@@ -147,26 +91,43 @@ struct medusa_kclass_s * med_get_kclass_by_pointer(struct medusa_kclass_s * ptr)
  * med_unregister_kclass soon.
  */
 
-int med_unlink_kclass(struct medusa_kclass_s * ptr)
+int _med_unlink_kclass(struct medusa_kclass_s *med_kclass)
 {
-	int retval = 0;
-	struct medusa_kclass_s * tmp;
+	struct medusa_kclass_s *pos, *prev;
+
+	if (med_kclass == kclasses) {
+		kclasses = med_kclass->next;
+		return 0;
+	}
+
+	for (prev = kclasses, pos = kclasses->next; pos != NULL; prev = pos, pos = pos->next) {
+		if (pos == med_kclass) {
+			prev->next = pos->next;
+			return 0;
+		}
+	}
+
+	med_pr_warn("Failed to unlink kclass '%s' - not found", med_kclass->name);
+	return -1;
+}
+
+int med_unlink_kclass(struct medusa_kclass_s *med_kclass)
+{
+	int retval;
 
 	MED_LOCK_W(registry_lock);
 	MED_LOCK_R(usecount_lock);
-	if (!ptr->use_count) {
-		if (ptr == kclasses)
-			kclasses = ptr->next;
-		else
-			for (tmp = kclasses; tmp; tmp = tmp->next)
-				if (tmp->next == ptr) {
-					tmp->next = tmp->next->next;
-					break;
-				}
-		/* TODO: verify whether we found it! */
```

Comment:
Congrats! After 20 years first Medusa TODO resolved!!!

Reply:
@matusjokay Actually, the reason behind the resolution (and the problem still exists in med_unregister_evtype) was not to resolve the TODO but unlinking of the kclass from linked list was not safe. If you check the old code, if you try to unlink the last node from _kclasses_, the _tmp->next = NULL_ and you will try to access _tmp->next->next_ (or the _next_ element of NULL node), which can result in undefined behavior.